### PR TITLE
Support processing v2

### PR DESCRIPTION
**Pull request by Siegrift**
Relates to https://github.com/api3dao/commons/issues/27

## Rationale

Read https://github.com/api3dao/commons/pull/32

## How to test

I used the development only invoke instructions to invoke signed API with pre/post processing configurations. Specifically:

Copy the pre-processing `packages/airnode-examples/integrations/coingecko-pre-processing/config.example.json` to `packages/airnode-node/config/config.json` and run the following:
```sh
AIRNODE_WALLET_PRIVATE_KEY=0x8127fa78936ebbd63b3f3e30272631d49abd6714c9349f430c383e52ac773cb9 yarn dev:http-signed-data-invoke -e 0x642598611f0dcbe389079bf555108513e3e8a15991887bb61126b7200f13c666 -p 0x3173690000000000000000000000000000000000000000000000000000000000636f696e49640000000000000000000000000000000000000000000000000000657468657265756d000000000000000000000000000000000000000000000000756e697854696d657374616d70000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000061a56980
```

Copy the post-processing `packages/airnode-examples/integrations/coingecko-post-processing/config.example.json` to `packages/airnode-node/config/config.json` and run the following:
```sh
AIRNODE_WALLET_PRIVATE_KEY=0x8127fa78936ebbd63b3f3e30272631d49abd6714c9349f430c383e52ac773cb9 yarn dev:http-signed-data-invoke -e 0x7ad7519b2b0ebd54c68356a314ffcacd7f897a08523e2ec9e0edd837fe995422 -p 0x3173730000000000000000000000000000000000000000000000000000000000636f696e49647300000000000000000000000000000000000000000000000000626974636f696e2c657468657265756d00000000000000000000000000000000767343757272656e6379000000000000000000000000000000000000000000007573640000000000000000000000000000000000000000000000000000000000
```

### CODE REVIEW

File: packages/airnode-node/src/api/index.ts [hunk 1]
```diff
@@ -300,28 +301,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
...
+  const { apiCallParameters: processedApiCallParameters } = await preProcessApiCallParameters(endpoint, parameters, {
     totalTimeoutMs: PROCESSING_TIMEOUT,
   });
 
   // skip API call if operation is undefined and fixedOperationParameters is empty array
   if (!endpoint.operation && isEmpty(endpoint.fixedOperationParameters)) {
-    // contents of preProcessingSpecifications or postProcessingSpecifications (or both) will simulate an API when API call is skipped
-    if (isEmpty(endpoint.preProcessingSpecifications) && isEmpty(endpoint.postProcessingSpecifications)) {
-      const message = `Failed to skip API call. Ensure at least one of 'preProcessingSpecifications' or 'postProcessingSpecifications' is defined and is not an empty array at ois '${payload.aggregatedApiCall.oisTitle}', endpoint '${payload.aggregatedApiCall.endpointName}'.`;
+    // Some processing needs to be defined.
```
Comment (Siegrift) [inline packages/airnode-node/src/api/index.ts:312]: https://github.com/api3dao/airnode/issues/1915 but I didn't want to do it now.

File: packages/airnode-node/src/api/index.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export function buildOptions(payload: ApiCallPayload): adapter.BuildRequestOptio
   const { type, config, aggregatedApiCall } = payload;
   const { endpointName, parameters, oisTitle } = aggregatedApiCall;
 
-  const ois = config.ois.find((o) => o.title === oisTitle)!;
+  const ois = config.ois.find((o) => o.title === oisTitle)! as OIS;
```
Comment (Siegrift) [inline packages/airnode-node/src/api/index.ts:30]: For some reason I was getting TS errors unless I did this. Not sure if it's caused by a different zod version, but I think someone added these in the past as well (probably for a similar reason).

File: packages/airnode-node/src/api/index.ts (hunk 1 above)
Comment (Siegrift) [reply packages/airnode-node/src/api/index.ts]: As @dcroote pointed out, the check can be removed because it's verified in the OIS.

File: packages/airnode-node/package.json [hunk 3]
```diff
@@ -29,8 +29,8 @@
     "@api3/airnode-protocol": "^0.13.0",
     "@api3/airnode-utilities": "^0.13.0",
     "@api3/airnode-validator": "^0.13.0",
-    "@api3/commons": "^0.3.0",
-    "@api3/ois": "2.2.1",
+    "@api3/commons": "file:../../../commons",
+    "@api3/ois": "file:../../../ois",
```
Comment (dcroote) [inline packages/airnode-node/package.json:33]: Want me to release OIS v2.3 now that https://github.com/api3dao/ois/pull/117 is merged?

Comment (Siegrift) [reply packages/airnode-node/package.json:33]: I've asked for npm access so I might do it myself.

File: packages/airnode-adapter/src/request-building/authentication.ts [hunk 4]
```diff
@@ -88,7 +88,19 @@ function getSchemeAuthentication(
...
+  // below be non-exhaustive and generate a TS error. This is a workaround.
+  type ApiSecuritySchemeType =
+    | 'apiKey'
+    | 'http'
+    | 'relayChainId'
+    | 'relayChainType'
+    | 'relayRequesterAddress'
+    | 'relaySponsorAddress'
+    | 'relaySponsorWalletAddress'
+    | 'relayRequestId';
```
Comment (dcroote) [inline packages/airnode-adapter/src/request-building/authentication.ts:101]: I am pretty sure there is something wrong with the `@api/ois` v2.3.0 release... 😕 

What I did:
- Confirmed the type is exported correctly in the latest airnode master (which uses @api3/ois v2.2.1)
- Checked out this PR, installed & built packages, and confirmed that the type is indeed now `any`
- Replaced `"@api3/ois": "2.3.0",` with `"@api3/ois": "file:../../../ois",` (which points to the latest master built locally) in `package.json` of `airnode-adapter`, reinstalled, and was able to delete this addition because the type is correct.
- Confirmed that `ApiSecurityScheme` in `dist/src/ois.d.ts` is `any` in `@api/ois` v2.3.0, but is correctly typed in that file when `@api3/ois` is built locally from the current master 

**Most concrete proof**:
- Published OIS v2.3.0 myself as https://www.npmjs.com/package/test-ois
- Replaced `@api3/ois` with `test-ois` and the typing is correct

Comment (Siegrift) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: > I am pretty sure there is something wrong with the @api/ois v2.3.0 release... 😕

That is pretty sad and also bad. I followed the release instructions (with the exception of pushing to main) so I don't think I made a mistake. I think that the files built on my machine somehow differ with yours...

Comment (Siegrift) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: > Published OIS v2.3.0 myself

Did you do it from the v2.3.0 commit?

Comment (dcroote) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: I published `2.3.0` of `test-ois` from the latest `main` (which includes 2 commits since v2.3.0). To see if those commits somehow had something to do with it, I just now checked out the v2.3.0 commit (https://github.com/api3dao/ois/commit/7d00c0a62bb5f28f5dc0b3601ecd7967773c2ea7), rebuilt, and [published that](https://www.npmjs.com/package/test-ois?activeTab=versions) as `2.3.0-alpha` of `test-ois`. I then used that as an Airnode dependency and same result- typing works.

So I'm not sure why your build appears to have differed, but I'd like to publish v2.3.1 of OIS to avoid hacking the typing here.

Comment (Siegrift) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: I troubleshoot this and I think I understand what happened. 

TLDR: We can just publish a 2.3.1 patch from main and it's going to work.

What happened:
1. I correctly published 2.3.0 but the typings are broken in Airnode (this PR)
2. In the meantime, you merged two dependency updates, including the Typescript dependency in OIS.
3. [Here](https://github.com/api3dao/airnode/pull/1914#discussion_r1408839509) you tried using the locally built OIS and the typings are correct (suggesting issue with 2.3.0 release).
4. [Here](https://github.com/api3dao/airnode/pull/1914#discussion_r1409553353) you've published `2.3.0-alpha` but you've actually not rebuilt that (you can verify that when you install the dependency and compare package.json with dist/package.json) so actually both `test-ois@2.3.0` and `test-ois@2.3.0-alpha` are "same" and both work.
5. I built the OIS myself, compared it deeply with `test-ois@2.3.0-alpha` and the content was the same (apart from discrepancies in the package.json).

Comment (dcroote) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: > TLDR: We can just publish a 2.3.1 patch from main and it's going to work.

Agree.

The rest is partially correct with a big caveat at the end. As a final step I first checked out https://github.com/api3dao/ois/commit/7d00c0a62bb5f28f5dc0b3601ecd7967773c2ea7, which is the v2.3.0 commit before the additional typescript and lodash dependabot commits (I didn't merge those btw). I then ran `yarn install && yarn build`, then released [test-ois@2.3.0-beta](https://www.npmjs.com/package/test-ois/v/2.3.0-beta?activeTab=versions). I then did a recursive diff within `dist` and it does differ from `test-ois@2.3.0`, which was built from master, as well as `test-ois@2.3.0-alpha`, which you correctly point out wasn't fully rebuilt before publishing and therefore is the same as `test-ois@2.3.0`.

However, probably the most important part is that `test-ois@2.3.0-beta` works in Airnode with the correct types, which points to something having gone wrong in your build and release of v2.3.0 (what exactly I have no clue).

Anyway, to me this is even more motivation for https://github.com/api3dao/ois/issues/122

Comment (dcroote) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: Just released v2.3.1 on npm and [GitHub](https://github.com/api3dao/ois/releases/tag/v2.3.1). As we expected, it looks to have solved this typing issue. Just remember to now bump & rerun the `update-ois-version.ts` script.

Comment (Siegrift) [reply packages/airnode-adapter/src/request-building/authentication.ts:101]: > However, probably the most important part is that test-ois@2.3.0-beta works in Airnode with the correct types, which points to something having gone wrong in your build and release of v2.3.0 (what exactly I have no clue).

You are right about this. I tried building the 2.3.0. locally and indeed it differs with what's published... This has to be some TS problem or maybe I didn't run `yarn` to install fresh dependencies 🤔 .

**Review by dcroote — COMMENTED** [review]
Are you sure your changes to the processing integrations work?

Neither the instructions you provided in your PR (which required a lot of work to even get to the error i.e. `httpSignedDataGateway` isn't enabled in `config.example.json` so `http-signed-data-invoke` fails) nor following the entire `coingecko-pre-processing` from scratch work for me.

Both fail with `TypeError: Cannot read properties of undefined (reading 'unixTimestamp')`

**Comment by Siegrift** [comment]
> Are you sure your changes to the processing integrations work?

You're right - sorry. I forgot to update the processing snippets after the renaming discussion we had at the docs PR. I also forgot to mention to copy the trigger in the config to the signed data gateway.

Please try again now.

File: packages/airnode-node/src/api/index.ts [hunk 5]
```diff
@@ -223,17 +223,18 @@ export async function processSuccessfulApiCall(
...
-    postProcessApiCallResponse(rawResponse.data, endpoint, aggregatedApiCall.parameters, {
+    postProcessResponse(rawResponse.data, endpoint, aggregatedApiCall.parameters, {
       totalTimeoutMs: PROCESSING_TIMEOUT,
     })
   );
   if (!goPostProcessApiSpecifications.success) {
     const log = logger.pend('ERROR', goPostProcessApiSpecifications.error.message);
     return [[log], { success: false, errorMessage: goPostProcessApiSpecifications.error.message }];
   }
+  const postProcessedValue = goPostProcessApiSpecifications.data.response;
```
[the remark this answers is not in the document]
Comment (dcroote) [reply packages/airnode-node/src/api/index.ts]: It doesn't look like you read this carefully

[the remark this answers is not in the document]
Comment (Siegrift) [reply packages/airnode-node/src/api/index.ts]: 🤦 Sorry, rough morning.

**Review by dcroote — APPROVED** [review]
Both pre- and post-processing integrations work now 👍
